Add Docker Registry support

drybjed opened the pull request:
Fixes #556

rei-ifesca committed:
modify gitlab config template to enable registry

rei-ifesca committed:
Add gitlab registry defaults variables and documentation

drybjed committed:
Merge remote-tracking branch 'rei-ifesca/gitlab_registry' into add-docker_registry

drybjed committed:
[debops.gitlab] Don't hard-depend on 'debops.pki'

The 'debops.pki' role does not have to be included as a dependency of
the 'debops.gitlab' role, because it is exepcted to be applied on the
hosts already.

Alternatively, it would be expected to be added to the role playbook as
a "soft" dependency instead of the role 'meta/main.yml' file as a "hard"
dependency. Any communication between roles can be faciliated via
Ansible local facts.

drybjed committed:
[debops.gitlab] Reorganize Registry variables

This patch reorganizes the default variables used to configure Gitlab
Container Registry service and cleans up their documentation. Some of
the variables required by the 'debops.docker_registry' role are exposed
via Ansible local facts.

drybjed committed:
Add 'debops.docker_registry' Ansible role

drybjed committed:
Improve integration of GitLab and Docker Registry

drybjed committed:
[debops.docker_registry] Fix problem with $PATH

The 'cron' service does not contain '/usr/local/bin/' directory in its
$PATH by default. To sidestep this in case the upstream
'docker-registry' binary is installed in '/usr/local/bin/', specify the
absolute path to the binary in the garbage collector script.

rei-ifesca commented on ansible/roles/debops.docker_registry/defaults/main.yml:
```diff
@@ -0,0 +1,629 @@
+# from GitLab Container Registry service.
+docker_registry__gitlab_token_realm_url: '{{ ansible_local.gitlab.registry_token_realm_url
+                                             if (ansible_local|d() and ansible_local.gitlab|d() and
+                                                 ansible_local.gitlab.registry_token_realm_url|d())
```
Apparently the indentation here is not correct.

drybjed replied:
I'm not sure what you mean, what would be the correct indentation?

rei-ifesca replied:
In the browser display it just looked "wrong". I think this review element can be  closed as solved.

rei-ifesca commented on ansible/roles/debops.docker_registry/defaults/main.yml:
```diff
@@ -0,0 +1,629 @@
+# configuration file.
+docker_registry__gitlab_token_issuer: '{{ ansible_local.gitlab.registry_token_issuer
+                                          if (ansible_local|d() and ansible_local.gitlab|d() and
+                                              ansible_local.gitlab.registry_token_issuer|d())
```
Apparently the indentation here is not correct.

drybjed replied:
I'm not sure what you mean, what would be the correct indentation?

rei-ifesca replied:
In the browser display it just looked "wrong". I think this review element can be  closed as solved.

rei-ifesca commented on ansible/roles/debops.docker_registry/defaults/main.yml:
```diff
@@ -0,0 +1,629 @@
+    auth_basic_name: '{{ docker_registry__basic_auth_name }}'
+
+    options: |
+      client_max_body_size 0;
```
Do you want to allow to upload images with insane sizes? 
What do you think? It is better to set a limit, lets say of 5GB or 10GB? 
It would at least reduce abuse, however this docker-registry will be mostly run as a private service...

drybjed replied:
Sure, I suppose a 10G limit by default should be fine, although I would expose it as a separate default variable for easy change. I'll add a quick commit in my next PR for this.

rei-ifesca replied:
A new variable sounds good.

rei-ifesca commented on ansible/roles/debops.docker_registry/defaults/main.yml:
```diff
@@ -0,0 +1,629 @@
+#
+# The TCP port on which the Docker Registry backend service listens for
+# connections.
+docker_registry__backend_port: '5070'
```
This is the port where the docker-registry daemon listens for HTTP connections, right?
So the same port number should be set in the gitlab defaults config file. There is the port set to 5005.

drybjed replied:
I tend to avoid using ports suggested by upstreams that are not officially registered - usually it's something like `4000`, `9000`, etc. The '5070' port seems to be available in the `/etc/services` database, so I went with it.

rei-ifesca replied:
Ok, that's an argument I did not looked at. The port number itself is for me irrelevant, I was just confused about the different numbers.

rei-ifesca commented on docs/ansible/roles/debops.gitlab/getting-started.rst:
```diff
@@ -54,6 +54,25 @@ on the host, for example by the :ref:`debops.redis_server` Ansible role.
+- Install GitLab, using facts from Docker Registry role
+- Update Docker Registry service with facts from GitLab role
+
+This should avoid issue with circular dependencies in the configuration of both
```
What do you think of using using the "/env" sub-role to configure both services first, like the nullmailer role:
` - role: debops.nullmailer/env`
Instead of running the playbooks again, although due to the idempotency of ansible nothing should break :D

drybjed replied:
Don't forget that the roles are designed to manage standalone services - Docker Registry and GitLab - separately. Ansible local facts allow for combining the two roles together while still working separately just fine.

Another problem is that both services require information from each other to perform correct integration. The order DR/G/DR seems to me to be a better choice, especially that the GitLab playbook runs significantly longer than Docker Registry.

I'm not sure what could be moved to an `/env` role - you still need to put the configuration file somewhere, for that you need to create UNIX accounts, and quickly any other options are exhausted. The current setup seems to be the most sane enough without merging the two roles together in some way and losing the ability to set them up separately.

rei-ifesca replied:
Understood. Lets keep the current setup.

rei-ifesca commented on ansible/roles/debops.gitlab/defaults/main.yml:
```diff
@@ -839,128 +839,124 @@ gitlab__piwik_url: ''
+gitlab__registry_host_port: '{{ ansible_local.docker_registry.host_port
+                                if (ansible_local|d() and ansible_local.docker_registry|d() and
+                                    ansible_local.docker_registry.host_port|d())
+                                else "5005" }}'
```
See my review regarding the registry listen port: https://github.com/debops/debops/pull/667/commits/83e42a69f4a97ed2a65215bbd9a9d28ef6920fad#diff-7f8f9d8ee0d63f8cbbdfe5b72cd893cbR188

drybjed replied:
If you don't use `debops.docker_registry` role to set up your Docker Registry, you probably have installed it manually or some other way, with default TCP ports. In that case, `debops.gitlab` role will try to connect to the service using the default ports suggested by upstream.

If you used the `debops.docker_registry`, the GitLab role should use the configuration defined by the Docker Registry role which could have used different TCP ports.

rei-ifesca replied:
Ok

rei-ifesca commented:
Good work! 👍 

What do you think, it is possible to move "old registry data" to the new registry data path and still have the old containers inside the new role?

Sadly I won't be able to test thid (huge) changes quickly, as our gitlab/registry is already running in production. We need to plan the migration

drybjed commented:
@rei-ifesca I found a couple of issues already, so the merge will probably happen during the weekend. I'll check if moving the data between directories changes anything, at least in the filesystem storage type.

drybjed committed:
[debops.docker_registry] Redesign token support

The support for token-based authentication in 'debops.docker_registry'
role was focused too much on support for GitLab Container Registry
service, therefore it was updated to be service-agnostic. The GitLab
support is enabled by default when the GitLab installation is detected
via Ansible local facts, but other "providers" can be used as well, with
easy switch between them.

drybjed committed:
[debops.docker_registry] Divert Debian Conffile

The 'docker-registry' Debian package contains a configuration file which
is marked as tracked by 'dpkg' (Conffile). The role will divert it to
ensure that there are no issues during upgrades.

drybjed committed:
[debops.docker_registry] Secure Registry config

Make sure tha only 'docker-registry' UNIX account can read the service
configuration file, since it may contain passwords.

drybjed committed:
[debops.docker_registry] Cleanup version string

drybjed committed:
[debops.docker_registry] Allow access to storage

The access to the Docker Registry storage directory is required by the
GitLab backup script to create backups of stored container images. Since
it was either allowing access to all UNIX accounts on the host, or
adding the 'git' UNIX account to the 'docker-registry' UNIX group,
I chose the former method as the default one.

drybjed committed:
[debops.gitlab] Allow access to PKI private keys

The 'git' UNIX account requires access to the private keys stored in the
DebOps PKI to load the private key used to sign authentication tokens
for Docker Registry service. This is done by adding the 'git' UNIX
account to the 'ssl-cert' UNIX group by default.

drybjed committed:
[debops.gitlab] Don't modify GitLab source code

The 'debops.gitlab' role modified the GitLab source code directly to
define a custom admin e-mail address as well as a random password. For
a long time now, GitLab supports providing the admin e-mail address via
an environment variable, and without a default password the application
will enforce setting a new 'root' account password on first use.

This patch removes custom tasks that modified the GitLab source code,
and provides the admin e-mail address via an environment variable to
leverage the new functionality.

drybjed committed:
[debops.gitlab] Don't generate Pages configuration

If the 'GitLab Pages' functionality is not enabled by the
'debops.gitlab' role, the 'debops.nginx' role will not generate a custom
configuration file with the server options for it. This is done to avoid
generating nginx configuration file with an empty 'server_name' which
could be mistakenly enabled in the webserver and break the existing
setup.

drybjed committed:
[debops.gitlab] Update upstream version to 11.7

drybjed committed:
[Travis] Add Ansible Galaxy webhook

drybjed committed:
[debops.gitlab] Update version in watch file

drybjed committed:
[Travis] Fix YAML indentation

drybjed committed:
[Travis] Don't notify Ansible Galaxy from Travis

This is not a good place for these kinds of notifications. This will be
performed from the GitHub repository instead.

drybjed commented:
@rei-ifesca I checked the data move between the directories, and it seems that it works as expected.

drybjed committed:
[debops.redis_server] Use redis.conf for passwords

The 'redis-password' script will now use the 'redis.conf' files which
have 'redis-auth' UNIX group. This allows the script to correctly look
up the password from an unprivileged account with access to the
'redis-auth' UNIX group.

drybjed committed:
[debops.docker_registry] Support older releases

The docker-registry versions before v2.7.1 do not support the
'--delete-untagged' parameter. The garbage collection script needs to
take it into account, otherwise the clean up process is not performed.

drybjed committed:
[debops.docker_registry] Clean Redis cache on GC

When Docker Registry performs garbage collection, it does not remove
existing data cached in Redis. This might cause issues with the Registry
when users try to push the same images a second time. Because of that,
when any garbage collection happens, the script will remove all Redis
keys related to Registry to avoid such issues.
